dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala

```diff
@@ -0,0 +1,332 @@
+import io.shiftleft.semanticcpg.language._
+
+import scala.collection.{Set, mutable}
+
+/** Creation of data dependence edges based on solution of the ReachingDefProblem.
+  */
+class DdgGenerator {
+
+  /** Once reaching definitions have been computed, we create a data dependence graph by checking which reaching
+    * definitions are relevant, meaning that a symbol is propagated that is used by the target node.
+    *
+    * @param dstGraph
+    *   the diff graph to add edges to
+    * @param problem
+    *   the reaching definition problem
+    * @param solution
+    *   the solution to `problem``
+    */
+  def addReachingDefEdges(
+    dstGraph: DiffGraphBuilder,
+    problem: DataFlowProblem[mutable.BitSet],
+    solution: Solution[mutable.BitSet]
+  ): Unit = {
+    implicit val implicitDst: DiffGraphBuilder = dstGraph
+    val numberToNode                           = problem.flowGraph.asInstanceOf[ReachingDefFlowGraph].numberToNode
+    val in                                     = solution.in
+    val gen = solution.problem.transferFunction
+      .asInstanceOf[ReachingDefTransferFunction]
+      .gen
+
+    val method        = problem.flowGraph.entryNode.asInstanceOf[Method]
+    val allNodes      = in.keys.toList
+    val usageAnalyzer = new UsageAnalyzer(problem, in)
+
+    /** Add an edge from the entry node to each node that does not have other incoming definitions.
+      */
+    def addEdgesFromEntryNode(): Unit = {
+      // Add edges from the entry node
+      allNodes
+        .filter(n => isDdgNode(n) && usageAnalyzer.usedIncomingDefs(n).isEmpty)
+        .foreach { node =>
+          addEdge(method, node)
+        }
+    }
+
+    /** Adds incoming edges to arguments of call sites, including edges between arguments of the same call site.
+      */
+    def addEdgesToCallSite(call: Call): Unit = {
+      // Edges between arguments of call sites
+      usageAnalyzer.usedIncomingDefs(call).foreach { case (use, ins) =>
+        ins.foreach { in =>
+          val inNode = numberToNode(in)
+          if (inNode != use) {
+            addEdge(inNode, use, nodeToEdgeLabel(inNode))
+          }
+        }
+      }
+
+      // For all calls, assume that input arguments
+      // taint corresponding output arguments
+      // and the return value. We filter invalid
+      // edges at query time (according to the given semantic).
+      usageAnalyzer.uses(call).foreach { use =>
+        gen(call).foreach { g =>
+          val genNode = numberToNode(g)
+          if (use != genNode && isDdgNode(use)) {
+            addEdge(use, genNode, nodeToEdgeLabel(use))
+          }
+        }
+      }
+
+      // Handle block arguments
+      call.argument.isBlock.foreach { block =>
+        block.astChildren.lastOption match {
+          case None => // Do nothing
+          case Some(node: Identifier) =>
+            val edgesToAdd = in(node).toList.flatMap { inDef =>
+              numberToNode(inDef) match {
```

Comment:
@DavidBakerEffendi this will crash with a match error if a non-identifier node is returned by the lookup. 
```
numberToNode.get(inDef).collect { case identifier: Identifier => identifier }
```
should cover all the cases.

Reply:
It would still need to be `case _ => None`, I think.

Reply:
I wondered about that too, but turns out it doesn't (at least if I understood the use case correctly):
```
ocular> val opt: Option[Object] = Some("Hello") 
opt: Option[Object] = Some(value = "Hello")

ocular> opt.collect { case x: Integer => x } 
res51: Option[Integer] = None

ocular> opt.collect { case x: String => x } 
res52: Option[String] = Some(value = "Hello")
```